Virtual Script Container (#39)
salesforce/lwc-rfcs: text/0113-secure-script-container.md

To ensure we can solve the problem while continuing to guarantee the benefits that are provided by the synthetic-shadow polyfill for all other components running on the main window, we are currently proposing the following characteristics for the virtual-script design:

* Create a JS sandbox before the Synthetic Shadow Polyfill is evaluated.
  * The sandbox is created using a same-domain iframe.

* Use this JS sandbox to virtualize the DOM APIs in order to prevent these APIs to be subject to the Synthetic Shadow Polyfill patches:
  * This can be done by caching references to the to-be-patched APIs inside the sandbox before the synthetic shadow is evaluated.

> * Evaluate one or more third party libraries inside this sandbox

Author: I think we may need a few more details here about how exactly this code gets evaluated. In the POC we are creating a `script` element and appending it to the `magicBody`. Does that actually get added to the `virtual-script`'s Shadow DOM? What if it fails?

* Run or include the third party library script inside this sandbox
  * This allows the code to run in a non-patched environment, while the rest of the document is running synthetic shadow
* Package the above functionality in a web component which semantically matches the `<script>` tag.

### Native Shadow

For applications which are using the native shadow, the underlying implementation will be different. There are two approaches we can choose from:

1. Virtualize every shadow DOM tree in the outer window into a single DOM structure which this container would see as the default DOM.
2. Implement a `document.querySelector` API which traverses the shadow DOM naturally.

### Globals

Some scripts define globals which are meant to be used by the rest of the application. Let's take this example of Google Analytics:

```html
<!-- Google Analytics -->
<script>
    window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
    ga('create', 'UA-XXXXX-Y', 'auto');
    ga('send', 'pageview');
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>
<!-- End Google Analytics -->
```

> If we included the above snippet inside our secure-script container, then `ga` will only be available to the *inner* `window`, i.e. the secure container's window. So `ga` will not be usable by anyone else in the entire application. We can however, allow for globals to be created inside our secure container, and redefined in any additional application sandboxes created by Locker (or the global outer window).

Author: I'm referring to the first inline script but we can assume the second one is referencing globals defined in the first one.

A syntax like the following could be provided:

```html

> <secure-script extra-globals="ga"></secure-script>

Author: > Just to confirm extra-globals="ga" means that the secure script can set the ga property on the global object. Right?

Yes, more specifically it means that any globals defined by a script in the secure container, are re-set in the outer window.

The secure script container does not have access to globals defined by other scripts. i.e. extra-globals does not define globals to keep in sync between the outer window and the secure script container. Rather it should copy any globals defined to the outer window.

So in your example it would log `undefined` even if `extra-globals` was set. I admit `extra-globals` may imply it's a pass-in.

> <secure-script extra-globals="ga"></secure-script>

Reviewer: > The secure script container does not have access to globals defined by other scripts

If it's the case, I don't see how the Google Analytics example works. The `analytics.js` does expect `ga` to be available on the `window` in order to run.

> <secure-script extra-globals="ga"></secure-script>

Author: The secure script would be the first one, as I said above, which defines the globals, like this:

```html
<!-- Google Analytics -->
<secure-script extra-globals="ga">
    window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
    ga('create', 'UA-XXXXX-Y', 'auto');
    ga('send', 'pageview');
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>
<!-- End Google Analytics -->
```

> <secure-script extra-globals="ga"></secure-script>

Reviewer: Shouldn't the analytics scripts be executed as well in a `secure-script` in order to work across shadows transparently?

> <secure-script extra-globals="ga"></secure-script>

Reviewer 2: yes, both should be executed under the `<virtual-script>`

> <secure-script extra-globals="ga"></secure-script>

Author: Yes, the point here was just to illustrate the global being referenced across script containers, both virtual and standard.

> <secure-script extra-globals="ga"></secure-script>

Reviewer 3: > The secure script container does not have access to globals defined by other scripts. i.e. extra-globals does not define globals to keep in sync between the outer window and the secure script container. Rather it should copy any globals defined to the outer window.

I think this should be defined in the spec, at least as a note.

> Yes, the point here was just to illustrate the global being referenced across script containers, both virtual and standard.

Does this negate @pmdartus point that due to not being able to access other globals that GA won't work?

> <secure-script extra-globals="ga"></secure-script>

Author: I think it seems like a good best practice but technically it should negate @pmdartus point since it would work, any globals created will be available in all sandboxes, and yeah I'll add that to the spec.

---
title: Secure Script Container
status: DRAFTED
created_at: 2020-07-15
pr: https://github.com/salesforce/lwc-rfcs/pull/38
champion: Caridy Patiño, Ted Conn
---

> # Secure Script Container for LWC Apps

Author: They are secure because they are running inside a secure javascript sandbox. I'm not tied to the name but `<lwc-script>` implies this is an LWC-only concept and it actually would work for any web component application.

> # Secure Script Container for LWC Apps

Reviewer 2: yeah, this is misleading for sure. We will change the name from `secure` to `virtual` since the goal here is not to secure anything, but to virtualize it.

> # Secure Script Container for LWC Apps

Reviewer 2: ok, changes are in, please, take another look.

## Summary

We need a way for application developers to integrate common third-party libraries into their LWC applications. These libraries, such as Google Analytics, are severely hindered by the shadow DOM which prevents global interaction with any element on the page.

By providing a secure container for scripts to run, see, and interact with all the shadow DOM trees as if it were the light DOM, application developers can continue to run their existing integrations until these libraries support shadow DOM semantics natively.

## Basic example

```html
<secure-script src="//cdn.optimizely.com/js/12345678.js"></secure-script>
```

> `secure-script` is a custom element which acts the same as the `script` tag, but any code which is evaluated inside this container is able to traverse the entire shadow tree using ligh DOM semantics. So if the above script runs `document.querySelectorAll('button')`, it will return all button elements regardless if they are in the shadow or not.

Reviewer 3: mis-spelling: `ligh` -> `light`

```html

> <secure-script>

Reviewer 4: I would love to see the equivalent imperative inline example, something like:

```js
SecureScript.evaluate(() => {});
```

Reviewer 4: That's a very fair point, we can do a follow up, however I would love to mention it here somewhere.

Reviewer 2: yes, we can mention it. I believe a static method on the VirtualScript web component class is sufficient, so you could import the class, and register yourself, extend it, or just call the static evaluator method. I will add a note for that to be a follow up proposal.

```

Inline scripting would also be supported, such as the above example which uses script injection to fetch its required resources.

## Motivation

Web component applications need the ability to integrate with third party libraries. Because
web components rely on the shadow DOM for encapsulation, this development paradigm does not work
with libraries who are expecting to interact with the application globally. Before there was
component encapsulation via shadow DOM it was possible for library authors to:

- Listen for events and know who fired them
- Define global styling rules that would get applied on every element
- Query the entire document for any element on the page

These libraries are typically added as global scripts to the document root rather than deeper 
integrations into the components themselves. When every component's root element is a shadow
root, there is no light DOM for them to work with, and their functionality is severely crippled.

Some examples of these libraries are:

- **Analytics tools**: Google Tag Manager / Google Analytics, Pendo
- **Personalization platforms**: Optimizely, Evergage, Google Optimize
- **Instrumentation tools**: New Relic, Sentry.io

> In addition, developers are finding that this model is not compatible with their **design systems** and **testing frameworks**.

Reviewer 4: I would argue that design systems will not fit into this model since they will still be bound to the shadow semantics

Author: I agree, this doesn't apply here.

The goal of this proposal is to provide a way for application developers to continue to use their existing integrations until library authors can support web component based applications.

## Design

`secure-script` will solve for the synthetic shadow and subsequently the native shadow.

### Synthetic Shadow
Many LWC applications apply the [synthetic-shadow polyfill](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/synthetic-shadow/). This polyfill emulates the native shadow behavior while still allowing global styles to cascade into the shadow trees. Synthetic shadow currently [patches](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/synthetic-shadow/src/polyfills/document-shadow) the `document.querySelector` and `document.querySelectorAll` APIs which is what prevents developers from penetrating the shadow DOM boundaries either in their LWC components or any other scripts added to the document. We only want to solve for the latter scenario.

> Ok, so what if we store a reference to the original, unpatched APIs before they get patched by our polyfill? And we let *only* our "secure scripts" access those APIs? That's the premise of this approach for synthetic shadow:

Reviewer 3: I think this paragraph could be re-worded. You ask two questions and then follow it with bullets. Would something like this work?

> To ensure we can solve the problem while continuing to guarantee the benefits that are provided by the synthetic-shadow polyfill for all other components, we are currently proposing the following characteristics for the secure-script design:

* Create a secure JS sandbox which doesn’t allow its object graph to be extended using [`Object.preventExtensions`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/preventExtensions)
  * The sandbox is created using a same-domain iframe provided by Locker v.Next

>   * This has to be done before the synthetic shadow is imported

Reviewer 2: because the goal is to make this to work with native shadow as well, we should not assume synthetic is present.

* Run or include the third party library script inside this sandbox
  * This allows the code to run in a non-patched environment, while the rest of the document is running synthetic shadow
* Package the above functionality in a web component which semantically matches the `<script>` tag.

### Native Shadow

> For applications which are using the native shadow, the underlying implementation will be different. There are two approaches we can choose from:

Reviewer 4: We should add way more detail here about how we are thinking on implementing this as we move forward

Author: I agree, I think it's important to understand that now, even if its implementation comes later.

Reviewer 2: will do in a bit!

Reviewer 2: added... please review

1. Virtualize every shadow DOM tree in the outer window into a single DOM structure which this container would see as the default DOM.
2. Implement a `document.querySelector` API which traverses the shadow DOM naturally.

> 

Reviewer 2: why do you think that native shadow is not possible? it is certainly more difficult to do, in fact it is the inverse implementation of the synthetic shadow polyfill, but we do have the expertise now, we can make it happen.

> 

Reviewer 5: Yeah, not saying that it is impossible, I trust that you can do the magic. But as you said it is going to be more difficult, so we could use the synthetic shadow dom until the tools are adapted, which I hope will happen (see, for example, what codecept.js is doing: https://codecept.io/shadow/). At least we can get Evergage to adopt a similar approach. For GTM, I'm seeing requests being sent at Google although I didn't see any roadmap yet.

> 

Reviewer 2: makes sense! I'm not that optimistic about tools being able to overcome the encapsulation paradigm dictated by shadow dom.

> 

Author: @caridy if that is the case then the shadow dom implementation will be much more important here. Do we need to flesh out any more details for how that will work?

        const src = this.textContent;
        if (src) {
            evaluate(src);
        }
    }
}

customElements.define('secure-script', SecureScript);
```

## Drawbacks

- We are potentially prolonging the life of our synthetic shadow dom polyfill, instead of encouraging community library authors to adopt shadow DOM semantics.
- The element may end up being anywhere from 5kb - 20kb.

## Alternatives

- **Allow shadow DOM to be toggled at a more granular level.** There is a forthcoming RFC for allowing component authors to decide if they want to opt in to shadow DOM or not. This would mean we can keep top-level elements and layout components in the light DOM, but base components or managed components can choose to safeguard their internals. Until we allow this, `secure-script` serves as a mitigation for application developers looking to add third party libraries to their apps.
- Do not apply the shadow DOM polyfill at all: this would allow component authors to violate shadow DOM encapsulation which is not an option.
- If we don't do this, then companies / websites building on our platform will not be able to integrate with common third party libraries.

## Adoption strategy

> TBD

Author: An application developer can choose to expose it or not, so I think it could be both. For Salesforce Communities / Commerce we would probably do the latter since we side on giving the admin the markup.

> TBD

Author: @gregwhitworth you said you had more opinions here?

> TBD

Reviewer 3: I'm hesitant to allow this to be a free-for-all due to this statement in your RFC

> We are potentially prolonging the life of our synthetic shadow dom polyfill, instead of encouraging community library authors to adopt Shadow DOM semantics.

I would prefer for us to leverage the top libs identified and give them access to this for Phase 1 and when the need for synthetic shadow is removed we can update the <n> libs to leverage Phase 2.

Is this a large request outside of these large products?

> TBD

Author: > we can update the libs

That part is throwing me off a bit. The only work on us will be to support a native shadow implementation of `virtual-script` but this will be transparent for libraries. I'm certainly misunderstanding something.

> TBD

Reviewer 3: Thanks, and sorry for the delayed response. We've since discussed this further but it comes down to my concern of compat risk, eg: who can use this and when? If we allow anyone to utilize it and run any code they want to bypass the shadow then we need to never close that hole. That puts a technical debt on the LWC team to maintain that and it may come with a whole host of other issues such as other teams that desire the encapsulation (for the very reason it exists) and them not being able to evolve due to this script. So ultimately I'd like to understand this plan further and given the business scenarios I'm aware of I think there can be a path forward that not only blocks us from a compat/tech debt perspective but also in the prodsec arena as well.

> TBD

Author: @gregwhitworth is it fair to say that the usage of `virtual-script` is outside of the scope of this RFC? I can certainly flesh that out in the "Rollout strategy". Something to the tune of "usage of virtual script containers require above average scrutiny because of a, b, c..."?

> TBD

Author: We discussed this offline and agreed to something like the above.

## How we teach this

TBD

## Unresolved questions

> 

Reviewer 2: the budget is around 5k. To support native shadow, it might increase a little bit more. @tedconn can you compute the minified/gzip size of the current demo for synthetic shadow? just to have a good reference.

> 

Reviewer 2: added the performence budget everywhere...

> - Do we want to support all the existing `script` tag attributes like `type`, `async` and `defer`?

Reviewer 3: are there any downsides to doing this? I'd prefer to not have to go implement any of what you just outlined and given that this is effectively a privileged script tag I can see authors adding in those attr thinking they'll work.

Author: Yeah. I was going to test if we can extend `HTMLScriptElement` and get some of this behavior for free.

Reviewer 4: I think the semantics of the `<secure-script>` will be by definition async, otherwise we will have to generate an equivalent imperative inline version so evaluation is sync.

Author: What is the motivation for matching imperative APIs, just programmatic convenience? Also there's nothing stopping us from having both, right?

```typescript
SecureScript.evaluateSync
SecureScript.evaluate
```

Reviewer: I do think like @diervo that `<secure-script>` should be async by nature but for different reasons. Preserving the default script semantic doesn't behavior would require to force HTML parsing to stop, and I don't feel that we should be in this business.

I would be supportive of adding the `defer` attribute.

Reviewer 4: Im not sure how we will be able to honor `defer` though

Reviewer 2: Ok, in today's demo, it supports everything supported by the original script tag, in fact, the way it works is that `<virtual-script>` becomes `<script>` with the same set of attributes and content inside the sandbox.

Now, many of the issues about scripts are for script tags parsed and executed as part of the html document, which is NEVER the case for `<virtual-script>` because it is always inserted dynamically inside the sandboxed iframe. In which case, browsers will do a good job no matter what to not block anything else. I don't think we need to complicate this more.

Reviewer: Since the elements are dynamically inserted into the document sandbox, the original script semantic is not preserved regarding timing and evaluation order. In the following example, we can ensure that `script1.js` is evaluated before `script2.js`.

```html
<script src="script1.js"></script>
<script src="script2.js"></script>
```

Now in the next example, we can guarantee the evaluation order since script processing is not done by the HTML parser. In some cases, `script2.js` might evaluate before `script1.js`.

```html
<virtual-script src="script1.js"></virtual-script>
<virtual-script src="script2.js"></virtual-script>
```

This is why I think one of the invariant here is that all `virtual-script`s should be considered `async` by nature.

Reviewer 2: I see... this is a very valid point, thanks for raising this issue @pmdartus , let me think more about it.

Author: The order is still preserved when multiple script tags are set as async, so forcing async doesn't solve the problem. For example:

```html
<script src="script1.js" async></script>
<script src="script2.js" async></script>
```

`script2` is always going to evaluate only after `script1` has been downloaded and evaluated.

@caridy can we insert each script into a single sandbox instead of having a separate sandbox per `virtual-script` tag?

Reviewer 2: Initially, that was my assumption, that they will be inserted on the same sandbox, we only built one... or a way to identify them somehow to be anchored to one or a handful of sandboxes.

Reviewer: > The order is still preserved when multiple script tags are set as async, so forcing async doesn't solve the problem.

@tedconn This is not the case, there is no execution order between async scripts.

If we evaluate the above snippets with `<virtual-script>` instead, then `ga` global value will only be available into the sandbox. So `ga` will not be usable by anyone else in the entire application. We can however, allow for globals to be exposed into the main window. A syntax like the following could be provided:

```html
<virtual-script extra-globals="ga"></virtual-script>
```

- `virtual-script` container would not have access to globals defined by other scripts in the outer window unless they were defined in a `virtual-script` using `extra-globals`.
- Globals defined in `extra-globals` will be redefined in any `virtual-script` containers.
- They will also be available to any Locker-created sandboxes

Note: these global values are controlled side-channels between the main window and the sandbox, and should be used with caution. For example, creating a component that depends on `ga` to be defined, is an anti-pattern.

### Supported Attributes

`crossorigin`: indicates that the browser should provide more information to window.onerror when handling CORS errors

`defer`: indicates that the script should be executed after the document has parsed (has no effect if `src` is absent)

> `integrity`: enables the browser to verify thata fetched resource is delivered without unexpected manipulation

Reviewer 6: ```suggestion
`integrity`: enables the browser to verify that a fetched resource is delivered without unexpected manipulation
```

`nomodule`: the script should not be executed in browsers that support ES2015 modules

> `nonce`: user provided nonce values for CSP will be supported

Reviewer 6: Will we be able to support CSP's `<hash-algorithm>-<base64-value>` to enable execution of inline virtual scripts?

Author: As long as virtual-script doesn't mutate the textContent of the tag then it shouldn't have an impact on that.

> `referrerpolicy`: indicates which referrer to send when fetching the focument

Reviewer 6: ```suggestion
`referrerpolicy`: indicates which referrer to send when fetching the script
```

A new element called `<virtual-script>` encapsulates the various mechanisms to evaluate the scripts in a virtual environment, which solves the problem when using synthetic shadow and subsequently, when using the native Shadow DOM. We can implement this solution in two phases:

### Phase I: Synthetic Shadow

Many LWC applications use the [synthetic-shadow polyfill](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/synthetic-shadow/). This polyfill emulates the native Shadow DOM behavior while still allowing global styles to cascade into the shadow trees. Synthetic Shadow patches many of the DOM APIs that allow you to interact and traverse the DOM tree, (e.g.: `document.querySelector` and `document.querySelectorAll`), these APIs will prevent developers from penetrating the Shadow DOM boundaries defined by their LWC components. In this phase, we want to solve the LWC + Synthetic Shadow scenario.

To ensure we can solve the problem while continuing to guarantee the benefits that are provided by the synthetic-shadow polyfill for all other components running on the main window, we are currently proposing the following characteristics for the virtual-script design:

> * Create a JS sandbox before the Synthetic Shadow Polyfill is evaluated.

Reviewer: Now that I think about it, what is the story around IE11? How does `<virtual-script>` behave when Locker can't be used?

Author: We aren't looking to support IE11 with `<virtual-script>`

web components rely on the Shadow DOM for encapsulation, this development paradigm does not work
with libraries who are expecting to interact with the application globally. Before there was
component encapsulation via Shadow DOM it was possible for library authors to:

- Listen for events and know who fired them
- Define global styling rules that would get applied on every element
- Query the entire document for any element on the page

These libraries are typically added as global scripts to the document root rather than deeper 
integrations into the components themselves. When every component in the application has a ShadowRoot attached to them, these libraries will not work as expected.

Some examples of these libraries are:

- **Analytics tools**: Google Tag Manager / Google Analytics, Pendo
- **Personalization platforms**: Optimizely, Evergage, Google Optimize
- **Instrumentation tools**: New Relic, Sentry.io

The goal of this proposal is to provide a way for application developers to continue to use their existing integrations until library authors support Shadow DOM integration.

## Design

> A new element called `<virtual-script>` encapsulates the various mechanism to evaluate the scripts in a virtual environment, which solves the problem when using synthetic shadow and subsequently, when using the native Shadow DOM. We can implement this solution in two phases:

Author: ```suggestion
A new element called `<virtual-script>` encapsulates the various mechanisms to evaluate the scripts in a virtual environment, which solves the problem when using synthetic shadow and subsequently, when using the native Shadow DOM. We can implement this solution in two phases:
```

### Phase I: Synthetic Shadow

Many LWC applications use the [synthetic-shadow polyfill](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/synthetic-shadow/). This polyfill emulates the native Shadow DOM behavior while still allowing global styles to cascade into the shadow trees. Synthetic Shadow patches many of the DOM APIs that allow you to interact and traverse the DOM tree, (e.g.: `document.querySelector` and `document.querySelectorAll`), these APIs will prevent developers from penetrating the Shadow DOM boundaries defined by their LWC components. In this phase, we want to solve the LWC + Synthetic Shadow scenario.

> To ensure we can solve the problem while continuing to guarantee the benefits that are provided by the synthetic-shadow polyfill for all other components running on the main window. We are currently proposing the following characteristics for the virtual-script design:

Author: ```suggestion
To ensure we can solve the problem while continuing to guarantee the benefits that are provided by the synthetic-shadow polyfill for all other components running on the main window, we are currently proposing the following characteristics for the virtual-script design:
```

Other considerations for Phase I:

 * due to the explicit order of evaluation required for this to function, we plan to only support one sandbox that can be shared by multiple virtual scripts.
 * budget is around 5kb (minify/gzip).
 * performance penalties will be added to the virtual scripts as they run inside a membrane:
   * evaluation budget: 30% slower than evaluating the same script in the main window.
   * operation budget: 50% slower than executing the same task is the main window.
   * 3rd party scripts are rarely part of the critical path for the application to function.

### Phase II: Native Shadow

For applications which are using the native shadow, the underlying implementation will be different, a more complicated one. We are currently proposing the following characteristics for the virtual-script design:

* During the evaluation of the virtual script component class, patch `Element.prototype.attachShadow`, to record any shadow created in the app.

> * When evaluating code via `<virtual-script>` element, create a new sandbox with patched DOM APIs to simulate a flatten tree.

Author: Why are we proposing to flatten instead of preserving the original DOM's hierarchy?

> This will require a query selector mechanism that can translate regular selectors into cross shadow selections

Eg: how will we reason with matching `element1 > element2` and *not* matching `element1 + element2`

Reviewer 2: We are talking about the native shadow here, so, there is no DOM hierarchy here that can work, we have to simulate it. I'm not sure I understand the question, but the point is, that in order for those 2 queries to function, specially when element1 is in one shadow, and element2 is in another shadow... how can you make that work? well, you have to reconstruct in memory a flatten representation of the intersection of those two shadow fragments, and do a query based on that... obviously that will be too slow, too complicated, but we can do something that represents that without having to construct it in memory. I have some ideas.

Author: Ok. When we say a flattened tree to me sounds like we would take something like this:

```html
<x-app>
  ▼ #shadow-root
    <x-child>
      ▼ #shadow-root
        <button />
    </x-child>
</x-app>
```

and flatten it to something like this:

```html
<x-app />
<x-child />
```

Reviewer 2: no, flatten tree reference to how browsers compute the UI elements, in this case, the flatten tree is:

```
x-app
   x-child
      button
```
